testing/e2e/e2e_test.go

```diff
@@ -0,0 +1,155 @@
+	path, err = filepath.Abs(filepath.Join("..", "..", "build", "e2e-certs"))
+	suite.Require().NoError(err)
+	suite.certPath = path
+	_, err = os.Stat(suite.certPath)
+	suite.Require().NoError(err)
+
+	// get env vars
+	v, ok := os.LookupEnv("ELASTICSEARCH_HOSTS")
+	suite.Require().True(ok, "expected ELASTICSEARCH_HOSTS to be defined")
+	suite.esHosts = v
+
+	v, ok = os.LookupEnv("ELASTICSEARCH_SERVICE_TOKEN")
+	suite.Require().True(ok, "expected ELASTICSEARCH_SERVICE_TOKEN to be defined")
+	suite.serviceToken = v
+
+	v, ok = os.LookupEnv("ELASTICSEARCH_USERNAME")
+	suite.Require().True(ok, "expected ELASTICSEARCH_USERNAME to be defined")
+	suite.elasticUser = v
+
+	v, ok = os.LookupEnv("ELASTICSEARCH_PASSWORD")
+	suite.Require().True(ok, "expected ELASTICSEARCH_PAASWORD to be defined")
+	suite.elasticPass = v
+
+	// create http.Client that trusts system CA and custom CA
+	cas, err := x509.SystemCertPool()
+	suite.Require().NoError(err)
+	p, err := os.ReadFile(filepath.Join(suite.certPath, "e2e-test-ca.crt"))
+	suite.Require().NoError(err)
+	ok = cas.AppendCertsFromPEM(p)
+	suite.Require().True(ok, "failed to add e2e-test-ca.crt to cert pool")
+
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{
+			RootCAs: cas,
+		},
+	}
+	suite.client = &http.Client{Transport: tr}
+}
+
+// SetupKibana will issue POST requests to Kibana's fleet API for setup
+func (suite *BaseE2ETestSuite) SetupKibana() {
+	req, err := http.NewRequest("POST", "http://localhost:5601/api/fleet/setup", nil)
+	suite.Require().NoError(err)
+	req.SetBasicAuth(suite.elasticUser, suite.elasticPass)
+	req.Header.Set("kbn-xsrf", "e2e-setup")
+
+	resp, err := suite.client.Do(req)
+	suite.Require().NoError(err)
+	resp.Body.Close()
+	suite.Require().Equal(http.StatusOK, resp.StatusCode, "unable to setup kibana fleet")
+
+	req, err = http.NewRequest("POST", "http://localhost:5601/api/fleet/agents/setup", nil)
+	suite.Require().NoError(err)
+	req.SetBasicAuth(suite.elasticUser, suite.elasticPass)
+	req.Header.Set("kbn-xsrf", "e2e-setup")
+
+	resp, err = suite.client.Do(req)
+	suite.Require().NoError(err)
+	resp.Body.Close()
+	suite.Require().Equal(http.StatusOK, resp.StatusCode, "unable to setup kibana fleet agents")
+}
+
+// IsFleetServerPortFree will check if port 8220 is free.
+// If it is in use it will poll every second for up to 30s for any change.
+func (suite *BaseE2ETestSuite) IsFleetServerPortFree() bool {
+	portFree := false
+	for i := 0; i < 30; i++ {
+		ln, err := net.Listen("tcp", ":8220")
+		if err == nil {
+			ln.Close()
+			portFree = true
+			break
+		}
+		time.Sleep(time.Second)
+	}
+	return portFree
+}
+
+// FleetServerStatusOK will poll fleet-server's status endpoint every second and return when it responds with a 200 status code
+// if the passed context terminates before a 200 is returned the current test will be marked as failed.
+func (suite *BaseE2ETestSuite) FleetServerStatusOK(ctx context.Context, url string) {
+	timer := time.NewTimer(time.Second)
```

Comment:
Nit, use a ticker so we don't need to remember to reset it on each retry?
```suggestion
	timer := time.NewTicker(time.Second)
	defer timer.Stop()
```

Reply:
using the timer approach ensures that we wait 1s in between attempts, if we use a ticker and for some reason the response takes a long time we would not wait as the next ticker would have triggered